File: 77. Combinations/Solution.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def backTrack(path, idx):
            if len(path) == k:
                result.append(path[:])
            
            for i in range(idx, n + 1):
                path.append(i)
                backTrack(path, i + 1)
                path.pop()
        
        result = []
        backTrack([], 1)
        return result
    

File: 77. Combinations/Solution.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> combine(int n, int k) {
        ArrayList<Integer> track = new ArrayList();
        backTrack(n, k, 1, track);
        return res;
    }
    
    public void backTrack(int n, int k, int start, ArrayList<Integer> track) {
        if (track.size() == k) {
            res.add(new ArrayList(track));
            return;
        }
        
        for (int i = start; i <= n; i++) {
            track.add(i);
            backTrack(n, k, i + 1, track);
            track.remove(track.size() - 1);
        }
    }
}